ethpm/contract.py

```diff
@@ -79,7 +66,8 @@ def link_bytecode(cls, attr_dict: Dict[str, str]) -> Type["LinkableContract"]:
         linked_class.bytecode_runtime = apply_all_link_references(
             linked_class.bytecode_runtime, linked_class.runtime_link_refs, attr_dict
         )
-        linked_class.is_bytecode_linked = True
+        linked_class.needs_bytecode_linking = False
+        cls.needs_bytecode_linking = True
```

Comment:
This doesn't seem right.

Instead of building `linked_class` and then mutating it with the newly linked bytecodes, can we instead link the values that need linking and gather up the newly linked `bytecode` and ` runtime_bytecode` values.  Then pass them into `factory`?

```python
runtime = apply_all_link_references(...)
bytecode = apply_all_link_references(...)
linked_class = cls.factory(cls.web3, bytecode_runtime=runtime, bytecode=bytecode)
return linked_class
```

We can add a sanity check that the bytecode is indeed linked before exiting by checking `linked_class.needs_bytecode_linking` which should be set to `False` if everything went well.